src/intrinsics.cpp

```diff
@@ -162,6 +279,127 @@ Symbol *lCreateISPCSymbolForLLVMIntrinsic(llvm::Function *func, SymbolTable *sym
+        TYs.push_back(type->LLVMType(g->ctx));
+    }
+    bool hasUnnamedType = false;
+    switch (ID) {
+    case ISPCIntrinsics::atomicrmw: {
+        retType = TYs[1];
+        name += origName + "." + lGetMangledTypeStr(TYs[1], hasUnnamedType);
+        break;
+    }
+    case ISPCIntrinsics::bitcast: {
+        Assert(TYs[0]->getPrimitiveSizeInBits() == TYs[1]->getPrimitiveSizeInBits());
+        retType = TYs[1];
+        name = ISPCIntrinsicsNames[(unsigned)ID];
+        name += "." + lGetMangledTypeStr(TYs[0], hasUnnamedType) + "." + lGetMangledTypeStr(TYs[1], hasUnnamedType);
+        break;
+    }
+    case ISPCIntrinsics::concat: {
+        assert(Type::Equal(argTypes[0], argTypes[1]));
+        retType = lGetDoubleWidthVectorType(TYs[0]);
+        name = ISPCIntrinsicsNames[(unsigned)ID];
+        name += "." + lGetMangledTypeStr(retType, hasUnnamedType) + "." + lGetMangledTypeStr(TYs[1], hasUnnamedType);
+        break;
+    }
+    case ISPCIntrinsics::cmpxchg: {
+        retType = TYs[1];
+        name += origName + "." + lGetMangledTypeStr(TYs[1], hasUnnamedType);
+        break;
+    }
+    case ISPCIntrinsics::extract: {
+        llvm::VectorType *vt = llvm::dyn_cast<llvm::VectorType>(TYs[0]);
+        Assert(vt);
+        retType = vt->getElementType();
+        name = ISPCIntrinsicsNames[(unsigned)ID];
+        name += "." + lGetMangledTypeStr(TYs[0], hasUnnamedType);
+        if (argTypes.size() == 3) {
+            name += "." + lGetMangledTypeStr(TYs[1], hasUnnamedType);
+        }
+        break;
+    }
+    case ISPCIntrinsics::fence: {
+        retType = llvm::Type::getVoidTy(*g->ctx);
+        name += origName;
+        break;
+    }
+    case ISPCIntrinsics::insert: {
+        llvm::VectorType *vt = llvm::dyn_cast<llvm::VectorType>(TYs[0]);
+        Assert(vt);
+        retType = vt;
+        name = ISPCIntrinsicsNames[(unsigned)ID];
+        name += "." + lGetMangledTypeStr(TYs[0], hasUnnamedType);
+        break;
+    }
+    case ISPCIntrinsics::packmask: {
+        retType = llvm::Type::getInt64Ty(*g->ctx);
+        name = ISPCIntrinsicsNames[(unsigned)ID];
+        name += "." + lGetMangledTypeStr(TYs[0], hasUnnamedType);
+        break;
+    }
+    case ISPCIntrinsics::select: {
+        retType = TYs[1];
+        name = ISPCIntrinsicsNames[(unsigned)ID];
+        name += "." + lGetMangledTypeStr(TYs[1], hasUnnamedType);
+        break;
+    }
+    case ISPCIntrinsics::stream_load: {
+        retType = TYs[1];
+        name = ISPCIntrinsicsNames[(unsigned)ID];
+        name += "." + lGetMangledTypeStr(TYs[1], hasUnnamedType);
+        break;
+    }
+    case ISPCIntrinsics::stream_store: {
+        retType = llvm::Type::getVoidTy(*g->ctx);
+        name = ISPCIntrinsicsNames[(unsigned)ID];
+        name += "." + lGetMangledTypeStr(TYs[1], hasUnnamedType);
+        break;
+    }
+    default: {
+        Error(SourcePos(), "Unknown ISPC intrinsic \"%s\".", ISPCIntrinsicsNames[(unsigned)ID]);
```

Comment:
This assumes ID is always valid but in case of unknown intrinsics it will be most likely invalid

Reply:
Done